File: codewave_py/codewave.py
import codewave_core.logger as logger

import codewave_core.util as util

import codewave_core.cmd_instance as cmd_instance
import codewave_core.cmd_finder as cmd_finder
import codewave_core.text_parser as text_parser
import codewave_core.closing_promp as closing_promp
import codewave_core.command as command

def init():
	command.initCmds()
	command.loadCmds()

class Codewave():
	def __init__(self,editor,parent = None, **keywords):
		self.editor,self.parent = editor,parent
		self.closingPromp = self.context = None
		self.nameSpaces = []
		self.vars = {}
		
		defaults = {
			'brakets' : '~~',
			'deco' : '~',
			'closeChar' : '/',
			'noExecuteChar' : '!',
			'carretChar' : '|',
			'checkCarret' : True
		}
		
		for key, val in defaults.items():
			if key in keywords:
				setattr(self,key,keywords[key])
			elif parent is not None :
				setattr(self,key,getattr(parent,key))
			else:
				setattr(self,key,val)
		
	def onActivationKey(self):
		logger.log('activation key')
		cmd = self.commandOnCursorPos()
		if cmd is not None :
			cmd.init()
			logger.log(cmd)
			cmd.execute()
		else:
			cpos = self.editor.getCursorPos()
			if cpos['start'] == cpos['end'] :
				self.addBrakets(cpos['start'],cpos['end'])
			else:
				self.promptClosingCmd(cpos['start'],cpos['end'])
	def commandOnCursorPos(self):
		cpos = self.editor.getCursorPos()
		return self.commandOnPos(cpos['end'])
	def commandOnPos(self,pos):
		if self.precededByBrakets(pos) and self.followedByBrakets(pos) and self.countPrevBraket(pos) % 2 == 1 :
			prev = pos-len(self.brakets)
			next = pos
		else :
			if self.precededByBrakets(pos) and self.countPrevBraket(pos) % 2 == 0:
				pos -= len(self.brakets)
			prev = self.findPrevBraket(pos)
			if prev is None :
				return None 
			next = self.findNextBraket(pos-1)
			if next is None or self.countPrevBraket(prev) % 2 != 0 :
				return None
		return cmd_instance.CmdInstance(self,prev,self.editor.textSubstr(prev,next+len(self.brakets)))
	def nextCmd(self,start = 0):
		pos = start
		beginning = None
		while True:
			f = self.findAnyNext(pos ,[self.brakets,"\n"])
			if f is None:
				break
			pos = f.pos + len(f.str)
			if f.str == self.brakets:
				if beginning is not None:
					return cmd_instance.CmdInstance(self,beginning,self.editor.textSubstr(beginning,f.pos+len(self.brakets)))
				else:
					beginning = f.pos
			else:
				beginning = None
		None
	def getEnclosingCmd(self,pos = 0):
		cpos = pos
		closingPrefix = self.brakets + self.closeChar
		while True:
			p = self.findNext(cpos,closingPrefix)
			if p is None:
				return None
			cmd = self.commandOnPos(p+len(closingPrefix))
			if cmd is not None:
				cpos = cmd.getEndPos()
				if cmd.pos < pos:
					return cmd
			else:
				cpos = p+len(losingPrefix)
	def precededByBrakets(self,pos):
		return self.editor.textSubstr(pos-len(self.brakets),pos) == self.brakets
	def followedByBrakets(self,pos):
		return self.editor.textSubstr(pos,pos+len(self.brakets)) == self.brakets
	def countPrevBraket(self,start):
		i = 0
		start = self.findPrevBraket(start)
		while start is not None :
			start = self.findPrevBraket(start)
			i += 1
		return i
	def isEndLine(self,pos):
		return self.editor.textSubstr(pos,pos+1) == "\n" or pos + 1 >= self.editor.textLen()
	def findLineStart(self,pos):
		p = self.findAnyNext(pos ,["\n"], -1)
		return p.pos+1 if p is not None else 0
	def findPrevBraket(self,start):
		return self.findNextBraket(start,-1)
	def findNextBraket(self,start,direction = 1):
		f = self.findAnyNext(start ,[self.brakets,"\n"], direction)
		if f is not None and f.str == self.brakets :
			return f.pos
	def findPrev(self,start,string):
		return self.findNext(start,string,-1)
	def findNext(self,start,string,direction = 1):
		f = self.findAnyNext(start ,[string], direction)
		if f is not None:
			return f.pos 
	def findAnyNext(self,start,strings,direction = 1):
		pos = start
		while True :
			if 0 > pos or pos >= self.editor.textLen() :
				return None
			for stri in strings :
			
				start, end = pos, pos + len(stri) * direction
				if end < start :
					start, end = end, start
				if stri == self.editor.textSubstr(start,end) :
					return util.StrPos(pos-len(stri) if direction < 0 else pos,stri)
			pos += direction
	def findMatchingPair(self,startPos,opening,closing,direction = 1):
		pos = startPos
		nested = 0
		while True:
			f = self.findAnyNext(pos,[closing,opening],direction)
			if f is None:
				break
			pos = f.pos + (len(f.str) if direction > 0 else 0)
			if f.str == (closing if direction > 0 else opening):
				if nested > 0:
					nested-=1
				else:
					return f
			else:
				nested+=1
		return None
	def addBrakets(self,start, end):
		if start == end :
			self.editor.insertTextAt(self.brakets+self.brakets,start)
		else :
			self.editor.insertTextAt(self.brakets,end)
			self.editor.insertTextAt(self.brakets,start)
		self.editor.setCursorPos(end+len(self.brakets))
	def promptClosingCmd(self,start, end):
		if self.closingPromp is not None:
			self.closingPromp.stop()
		self.closingPromp = closing_promp.ClosingPromp(self,start, end).begin()
	def parseAll(self,recursive = True):
		pos = 0
		while True:
			cmd = self.nextCmd(pos)
			if cmd is None:
				break
			pos = cmd.getEndPos()
			self.editor.setCursorPos(pos)
			if recursive and cmd.content is not None :
				parser = Codewave(text_parser.TextParser(cmd.content),parent=self)
				cmd.content = parser.parseAll()
			if cmd.init().execute() is not None:
				if cmd.replaceEnd is not None:
					pos = cmd.replaceEnd
				else:
					pos = self.editor.getCursorPos().end
		return self.getText()
	def getText(self):
		return self.editor.text
	def getNameSpaces(self):
		npcs = set(['core']).union(self.nameSpaces)
		if self.parent is not None:
			npcs = npcs.union(self.parent.getNameSpaces())
		if self.context is not None:
			if self.context.finder is not None:
				npcs = npcs.union(self.context.finder.namespaces)
			npcs = npcs.union([self.context.cmd.fullName])
		return list(npcs)
	def addNameSpace(self,name):
		self.nameSpaces.append(name)
	def removeNameSpace(self,name):
		self.nameSpaces = [ n for n in self.nameSpaces if n != name]
	def getCmd(self,cmdName,nameSpaces = []) :
		finder = self.getFinder(cmdName,nameSpaces)
		found = finder.find()
		return found
	def getFinder(self,cmdName,nameSpaces = []) :
		return cmd_finder.CmdFinder(cmdName,self.getNameSpaces() + nameSpaces,
			useDetectors = self.isRoot(),
			codewave = self
		)
	def isRoot(self):
		return self.parent is None and (self.context is None or self.context.finder is None)
	def getCommentChar(self):
		return '<!-- %s -->'
	def wrapComment(self,str):
		cc = self.getCommentChar()
		if '%s' in cc :
			return cc.replace('%s',str)
		else:
			return cc + ' ' + str + ' ' + cc
	def wrapCommentLeft(self,str = ''):
		cc = self.getCommentChar()
		i = cc.index('%s') if '%s' in cc else None
		if i is not None:
			return cc[0:i] + str
		else:
			return cc + ' ' + str
	def wrapCommentRight(self,str = ''):
		cc = self.getCommentChar()
		i = cc.index('%s') if '%s' in cc else None
		if i is not None:
			return str + cc[i+2:]
		else:
			return str + ' ' + cc
	def removeCarret(self,txt):
		return txt.replace(self.carretChar+self.carretChar, '[[[[quoted_carret]]]]') \
							.replace(self.carretChar, '') \
							.replace('[[[[quoted_carret]]]]', self.carretChar)
	def getCarretPos(self,txt):
		txt = txt.replace(self.carretChar+self.carretChar, ' ')
		if self.carretChar in txt :
			return txt.index(self.carretChar)

File: lib/Codewave.js

const Process = require('./Process').Process

const Context = require('./Context').Context

const PositionedCmdInstance = require('./PositionedCmdInstance').PositionedCmdInstance

const TextParser = require('./TextParser').TextParser

const Command = require('./Command').Command

const Logger = require('./Logger').Logger

const PosCollection = require('./positioning/PosCollection').PosCollection

const StringHelper = require('./helpers/StringHelper').StringHelper

const ClosingPromp = require('./ClosingPromp').ClosingPromp

const defaultOptions = {
  brakets: '~~',
  deco: '~',
  closeChar: '/',
  noExecuteChar: '!',
  carretChar: '|',
  checkCarret: true,
  inInstance: null
}

var Codewave = class Codewave {
  constructor (editor, options = {}) {
    this.editor = editor
    Codewave.init()
    this.marker = '[[[[codewave_marquer]]]]'
    this.vars = {}
    const defaults = Codewave.defaultOptions
    this.parent = options.parent
    this.nested = this.parent != null ? this.parent.nested + 1 : 0

    Object.keys(defaults).forEach((key) => {
      if (key in options) {
        this[key] = options[key]
      } else if (this.parent != null && key !== 'parent') {
        this[key] = this.parent[key]
      } else {
        this[key] = defaults[key]
      }
    })

    if (this.editor != null) {
      this.editor.bindedTo(this)
    }

    this.context = new Context(this)

    if (this.inInstance != null) {
      this.context.parent = this.inInstance.context
    }

    this.logger = new Logger()
  }

  onActivationKey () {
    this.process = new Process()
    this.logger.log('activation key')
    return this.runAtCursorPos().then(() => {
      this.process = null
    })
  }

  runAtCursorPos () {
    if (this.editor.allowMultiSelection()) {
      return this.runAtMultiPos(this.editor.getMultiSel())
    } else {
      return this.runAtPos(this.editor.getCursorPos())
    }
  }

  runAtPos (pos) {
    if (pos == null) {
      throw new Error('Cursor Position is empty')
    }

    return this.runAtMultiPos([pos])
  }

  runAtMultiPos (multiPos) {
    return Promise.resolve().then(() => {
      if (multiPos.length > 0) {
        const cmd = this.commandOnPos(multiPos[0].end)

        if (cmd != null) {
          if (multiPos.length > 1) {
            cmd.setMultiPos(multiPos)
          }

          cmd.init()
          this.logger.log(cmd)
          return cmd.execute()
        } else {
          if (multiPos[0].start === multiPos[0].end) {
            return this.addBrakets(multiPos)
          } else {
            return this.promptClosingCmd(multiPos)
          }
        }
      }
    })
  }

  commandOnPos (pos) {
    let next, prev

    if (this.precededByBrakets(pos) && this.followedByBrakets(pos) && this.countPrevBraket(pos) % 2 === 1) {
      prev = pos - this.brakets.length
      next = pos
    } else {
      if (this.precededByBrakets(pos) && this.countPrevBraket(pos) % 2 === 0) {
        pos -= this.brakets.length
      }

      prev = this.findPrevBraket(pos)

      if (prev == null) {
        return null
      }

      next = this.findNextBraket(pos - 1)

      if (next == null || this.countPrevBraket(prev) % 2 !== 0) {
        return null
      }
    }

    return new PositionedCmdInstance(this, prev, this.editor.textSubstr(prev, next + this.brakets.length))
  }

  nextCmd (start = 0) {
    var beginning, f, pos
    pos = start

    // eslint-disable-next-line no-cond-assign
    while (f = this.findAnyNext(pos, [this.brakets, '\n'])) {
      pos = f.pos + f.str.length

      if (f.str === this.brakets) {
        if (typeof beginning !== 'undefined' && beginning !== null) {
          return new PositionedCmdInstance(this, beginning, this.editor.textSubstr(beginning, f.pos + this.brakets.length))
        } else {
          beginning = f.pos
        }
      } else {
        beginning = null
      }
    }

    return null
  }

  getEnclosingCmd (pos = 0) {
    var closingPrefix, cpos, p
    cpos = pos
    closingPrefix = this.brakets + this.closeChar

    while ((p = this.findNext(cpos, closingPrefix)) != null) {
      const cmd = this.commandOnPos(p + closingPrefix.length)
      if (cmd) {
        cpos = cmd.getEndPos()

        if (cmd.pos < pos) {
          return cmd
        }
      } else {
        cpos = p + closingPrefix.length
      }
    }

    return null
  }

  precededByBrakets (pos) {
    return this.editor.textSubstr(pos - this.brakets.length, pos) === this.brakets
  }

  followedByBrakets (pos) {
    return this.editor.textSubstr(pos, pos + this.brakets.length) === this.brakets
  }

  countPrevBraket (start) {
    var i
    i = 0

    while ((start = this.findPrevBraket(start)) != null) {
      i++
    }

    return i
  }

  isEndLine (pos) {
    return this.editor.textSubstr(pos, pos + 1) === '\n' || pos + 1 >= this.editor.textLen()
  }

  findPrevBraket (start) {
    return this.findNextBraket(start, -1)
  }

  findNextBraket (start, direction = 1) {
    const f = this.findAnyNext(start, [this.brakets, '\n'], direction)

    if (f && f.str === this.brakets) {
      return f.pos
    }
  }

  findPrev (start, string) {
    return this.findNext(start, string, -1)
  }

  findNext (start, string, direction = 1) {
    const f = this.findAnyNext(start, [string], direction)

    if (f) {
      return f.pos
    }
  }

  findAnyNext (start, strings, direction = 1) {
    return this.editor.findAnyNext(start, strings, direction)
  }

  findMatchingPair (startPos, opening, closing, direction = 1) {
    var f, nested, pos
    pos = startPos
    nested = 0

    // eslint-disable-next-line no-cond-assign
    while (f = this.findAnyNext(pos, [closing, opening], direction)) {
      pos = f.pos + (direction > 0 ? f.str.length : 0)

      if (f.str === (direction > 0 ? closing : opening)) {
        if (nested > 0) {
          nested--
        } else {
          return f
        }
      } else {
        nested++
      }
    }

    return null
  }

  addBrakets (pos) {
    pos = new PosCollection(pos)
    const replacements = pos.wrap(this.brakets, this.brakets).map(function (r) {
      return r.selectContent()
    })
    return this.editor.applyReplacements(replacements)
  }

  promptClosingCmd (selections) {
    if (this.closingPromp != null) {
      this.closingPromp.stop()
    }

    this.closingPromp = ClosingPromp.newFor(this, selections).begin()
  }

  newInstance (editor, options) {
    return new Codewave(editor, options)
  }

  parseAll (recursive = true) {
    var cmd, parser, pos, res

    if (this.nested > 100) {
      throw new Error('Infinite parsing Recursion')
    }

    pos = 0

    // eslint-disable-next-line no-cond-assign
    while (cmd = this.nextCmd(pos)) {
      pos = cmd.getEndPos()
      this.editor.setCursorPos(pos) // console.log(cmd)

      cmd.init()

      if (recursive && cmd.content != null && (cmd.getCmd() == null || !cmd.getOption('preventParseAll'))) {
        parser = new Codewave(new TextParser(cmd.content), {
          parent: this
        })
        cmd.content = parser.parseAll()
      }

      res = cmd.execute()

      if (res != null) {
        if (res.then != null) {
          throw new Error('Async nested commands are not supported')
        }

        if (cmd.replaceEnd != null) {
          pos = cmd.replaceEnd
        } else {
          pos = this.editor.getCursorPos().end
        }
      }
    }

    return this.getText()
  }

  getText () {
    return this.editor.text()
  }

  isRoot () {
    return this.parent == null && (this.inInstance == null || this.inInstance.finder == null)
  }

  getRoot () {
    if (this.isRoot()) {
      return this
    } else if (this.parent != null) {
      return this.parent.getRoot()
    } else if (this.inInstance != null) {
      return this.inInstance.codewave.getRoot()
    }
  }

  getFileSystem () {
    if (this.editor.fileSystem) {
      return this.editor.fileSystem
    } else if (this.isRoot()) {
      return null
    } else if (this.parent != null) {
      return this.parent.getRoot()
    } else if (this.inInstance != null) {
      return this.inInstance.codewave.getRoot()
    }
  }

  removeCarret (txt) {
    return StringHelper.removeCarret(txt, this.carretChar)
  }

  getCarretPos (txt) {
    return StringHelper.getCarretPos(txt, this.carretChar)
  }

  regMarker (flags = 'g') {
    return new RegExp(StringHelper.escapeRegExp(this.marker), flags)
  }

  removeMarkers (text) {
    return text.replace(this.regMarker(), '')
  }

  static init () {
    if (!this.inited) {
      this.inited = true

      Command.initCmds()

      return Command.loadCmds()
    }
  }
}

Codewave.inited = false
Codewave.defaultOptions = defaultOptions

exports.Codewave = Codewave
